File: src/main/java/ml/duncte123/skybot/commands/music/ListCommand.java
package ml.duncte123.skybot.commands.music;

import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import ml.duncte123.skybot.audio.GuildMusicManager;
import ml.duncte123.skybot.audio.TrackScheduler;
import ml.duncte123.skybot.objects.command.Command;
import ml.duncte123.skybot.utils.AirUtils;
import ml.duncte123.skybot.utils.AudioUtils;
import ml.duncte123.skybot.utils.EmbedUtils;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent;

import java.util.Queue;

public class ListCommand extends Command {

    /**
     * This is the executeCommand of the command, the thing you want the command to to needs to be in here
     * @param args The command agruments
     * @param event a instance of {@link net.dv8tion.jda.core.events.message.guild.GuildMessageReceivedEvent GuildMessageReceivedEvent}
     */
    @Override
    public void executeCommand(String[] args, GuildMessageReceivedEvent event) {
        AudioUtils au = AirUtils.au;

        Guild guild = event.getGuild();
        GuildMusicManager mng = au.getMusicManager(guild);
        TrackScheduler scheduler = mng.scheduler;

        Queue<AudioTrack> queue = scheduler.queue;
        synchronized (queue) {
            if (queue.isEmpty()) {
                sendEmbed(EmbedUtils.embedField(au.embedTitle, "The queue is currently empty!"), event);
            } else {
                int trackCount = 0;
                long queueLength = 0;
                StringBuilder sb = new StringBuilder();
                sb.append("Current Queue: Entries: ").append(queue.size()).append("\n");
                for (AudioTrack track : queue) {
                    queueLength += track.getDuration();
                    if (trackCount < 10) {
                        sb.append("`[").append(AudioUtils.getTimestamp(track.getDuration())).append("]` ");
                        sb.append(track.getInfo().title).append("\n");
                        trackCount++;
                    }
                }
                sb.append("\n").append("Total Queue Time Length: ").append(AudioUtils.getTimestamp(queueLength));
               sendEmbed(EmbedUtils.embedField(au.embedTitle, sb.toString()), event);
            }
        }
    }

    /**
     * The usage instructions of the command
     * @return a String
     */
    @Override
    public String help() {
        // TODO Auto-generated method stub
        return "shows the current queue";
    }

    @Override
    public String getName() {
        return "list";
    }

    @Override
    public String[] getAliases() {
        return new String[]{"queue"};
    }
}

File: src/main/kotlin/ml/duncte123/skybot/commands/music/ListCommand.kt
/*
 * Skybot, a multipurpose discord bot
 *      Copyright (C) 2017 - 2019  Duncan "duncte123" Sterken & Ramid "ramidzkh" Khan
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package ml.duncte123.skybot.commands.music

import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import me.duncte123.botcommons.messaging.EmbedUtils
import me.duncte123.botcommons.messaging.MessageUtils.sendEmbed
import ml.duncte123.skybot.Author
import ml.duncte123.skybot.objects.command.CommandContext
import ml.duncte123.skybot.objects.command.MusicCommand
import ml.duncte123.skybot.utils.AudioUtils
import ml.duncte123.skybot.utils.AudioUtils.getTimestamp
import org.apache.commons.lang3.StringUtils
import java.util.*

@Author(nickname = "Sanduhr32", author = "Maurice R S")
class ListCommand : MusicCommand() {

    init {
        this.name = "list"
        this.aliases = arrayOf("queue", "q")
        this.helpFunction = {_,_ -> "Shows the current queue"}
    }

    override fun run(ctx: CommandContext) {
        val event = ctx.event
        val mng = getMusicManager(event.guild, ctx.audioUtils)
        val scheduler = mng.scheduler
        val queue: Queue<AudioTrack> = scheduler.queue

        synchronized(queue) {
            if (queue.isEmpty()) {
                sendEmbed(event, EmbedUtils.embedField(AudioUtils.EMBED_TITLE, "The queue is currently empty!"))
            } else {
                val queueLength = queue.map { it.duration }.sum()
                val maxTracks = 10
                val queueText = buildString {
                    appendln("Current Queue: Entries: ${queue.size}")

                    for ((index, track) in queue.withIndex()) {
                        if (index == maxTracks) {
                            break
                        }

                        appendln(StringUtils.abbreviate("`[${getTimestamp(track.duration)}]` ${track.info.title}", 40))
                    }

                    appendln("Total Queue Time Length: ${getTimestamp(queueLength)}")
                    appendln("Hint: Use `${ctx.prefix}save` to save the current queue to a file that you can re-import")
                }

                sendEmbed(event, EmbedUtils.embedField(AudioUtils.EMBED_TITLE, queueText))
            }
        }
    }

}
